src/Compiler/test/Microsoft.NET.Sdk.Razor.SourceGenerators.Tests/RazorSourceGeneratorTestsBase.cs

```diff
@@ -80,15 +90,79 @@ protected static async ValueTask<(GeneratorDriver, ImmutableArray<AdditionalText
 
     protected static GeneratorRunResult RunGenerator(Compilation compilation, ref GeneratorDriver driver, params Action<Diagnostic>[] expectedDiagnostics)
     {
-        driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out compilation, out _);
+        return RunGenerator(compilation, ref driver, out _, expectedDiagnostics);
+    }
+
+    protected static GeneratorRunResult RunGenerator(Compilation compilation, ref GeneratorDriver driver, out Compilation outputCompilation, params Action<Diagnostic>[] expectedDiagnostics)
+    {
+        driver = driver.RunGeneratorsAndUpdateCompilation(compilation, out outputCompilation, out _);
 
-        var actualDiagnostics = compilation.GetDiagnostics().Where(d => d.Severity != DiagnosticSeverity.Hidden);
+        var actualDiagnostics = outputCompilation.GetDiagnostics().Where(d => d.Severity != DiagnosticSeverity.Hidden);
         Assert.Collection(actualDiagnostics, expectedDiagnostics);
 
         var result = driver.GetRunResult();
         return result.Results.Single();
     }
 
+    protected static async Task<string> RenderRazorPageAsync(Compilation compilation, string name)
+    {
+        // Load the compiled DLL.
+        Assembly assembly;
+        using (var peStream = new MemoryStream())
+        {
+            var emitResult = compilation.Emit(peStream);
+            Assert.True(emitResult.Success, string.Join(Environment.NewLine, emitResult.Diagnostics));
+            assembly = Assembly.Load(peStream.ToArray());
```

Comment:
Do we need to worry about isolating these loads? I'm a bit concerned that about potential ambient state if we load thousands of test dlls into our test process.

Reply:
FWIW, we already do this, since I basically copied that code from here:

Reply:
I see. That's horrifying.

Reply:
I concur ... that is a bit horrifying. This works in Roslyn but because we have a ton of infra in the background to properly isolate the loads and make sure they don't conflict with each other. I think we'll have to eventually port that logic over here. It should be a bit simpler cause we only have to deal with `AssemblyLoadContext` where as Roslyn has to swap between that and `AppDomain`. 

What all do we do with these assemblies? Do we end up executing code from them or just poke around the metadata?

Reply:
> What all do we do with these assemblies? Do we end up executing code from them or just poke around the metadata?

It seems the existing code mainly pokes around the metadata, although sometimes it creates instances of types from the assembly to do that, e.g.:

https://github.com/dotnet/razor/blob/789efc684bb1b4d92c27240def43e7891d7d3276/src/Compiler/test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/RazorIntegrationTestBase.cs#L379-L390

https://github.com/dotnet/razor/blob/789efc684bb1b4d92c27240def43e7891d7d3276/src/Compiler/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentDirectiveIntegrationTest.cs#L30-L38

My tests in this PR go further, they also execute the code.